keps/0035-json-schema-export.md

```diff
@@ -0,0 +1,296 @@
+
+#### Additional Attributes
+
+JSON-Schema already provides good support for generated UIs, but it can be enhanced. Additionally, we do not want to adopt the full nested structure with alternatives and references yet, there is a different KEP for that. The new attributes should provide a step in the direction to generate a good UI from it
+
+All of the new attributes introduced below are fully optional.
+
+`group` 
+An attribute of type "string". This attribute describes a top-level group used for this parameter. The value must not contain the character `/` as this is reserved for future support of nested groups. At the moment, nested groups are *not* supported.
+
+`advanced` 
+An attribute of type "boolean" that indicates that this parameter is for advanced configuration and can be hidden by default. Defaults to "false". Parameters that have "advanced=true" *must* have a default value, as they might not be visible immediately.
+
+`hint`
+An attribute of type "string" that should provide a short description about the syntax of this parameter. Usually shown below the actual input element.
+
+`enum` 
+An array of values that are allowed to be used. Each entry must be of the type defined by the `type` attribute. If a default is specified, it must be listed as well.
+
+
+Example:
+
+```yaml
+  - name: NODE_DOCKER_IMAGE_PULL_POLICY
+    title: "Docker Image Pull Policy"
+    hint: "Cassandra node Docker image pull policy"
+    description: "The Docker image pull policy. Can be changed, for example if the image is pre-loaded on the Kubernetes nodes."
+    default: "Always"
+    type: "string"
+    enum: 
+      - "Always"
+      - "Never"
+      - "IfNotPresent"
+    group: "Node"
+    advanced: true
+    required: true
+```
+
+#### Groups
+
+Add a top-level element `groups` that describes the groups used:
+
+```
+apiVersion: kudo.dev/v1beta1
+groups:
+  - name: "Node"
+    title: "Node Configuration"
+    description: "This section contains parameters that are related to the Node"
+parameters:
+...
+```
+
+This `groups` element is fully optional and does not affect any behaviour of KUDO. We may decide to not add this to the OperatorVersion CRD but to keep it only in the `params.yaml` package format, that is an implementation detail. The data provided in this element can be used by the JSON-Schema export to fill in the metadata (title and description) of top level groups.
+
+### JSON-Schema export
+
+Extend the `kudo package list parameters` command with the `--output` flag:
+
+Examples:
+
+`kudo package list parameters`  
+Outputs a human readable list as before
+
+`kudo package list parameters --output yaml`  
+Outputs the `params.yaml` file as is
+
+`kudo package list parameters --output json`  
+Outputs the `params.yaml` file in json format
+
+`kudo package list parameters --output json-schema`  
+Outputs the parameters as JSON-Schema
+
+`kudo package list parameters --output json-schema-yaml`  
```

Comment:
Hm, I'm not sure about this. `--output` is about the output format. If we say that YAML is one and JSON too, can we really say that `json-schema` is one too? And `json-schema-yaml` sounds strange to me (what is it now? JSON? YAML?)

Reply:
I've had another approach here: https://github.com/kudobuilder/kudo/pull/1698/commits/f9e32949cb58808f15b101a7dfb4735a30d793f9#diff-78066f3086b3541869b714f9b97d9814R143

Where I used `--output` *and* `--format`, but @alenkacz convinced me that using only `--output` would be ok. We copied that parameter from kubectl, which also uses multiple variants ( i.e. `-o wide`, etc.)

As for `--output json-schema-yaml` I agree that it's a bit cumbersome, but better than `--output json-schema-as-yaml` and `--output yaml-schema` is really wrong in my eyes.

I feel this is the most pragmatic approach